File: sdk/typescript/src/api/api-keys.ts
import type { HttpClient, RequestOptions } from "../http.js";

export type ApiKeyScope = "read" | "write" | "inference";

export interface CreateApiKeyBody {
  name: string;
  scopes?: ApiKeyScope[];
  allowedIps?: string[];
  expiresAt?: string;
}

export class ApiKeysApi {
  constructor(private readonly http: HttpClient) {}

  listApiKeys<T = unknown>(options?: RequestOptions): Promise<T> {
    return this.http.get<T>("/api-keys", options);
  }

  createApiKey<T = unknown>(
    body: CreateApiKeyBody,
    options?: RequestOptions,
  ): Promise<T> {
    return this.http.post<T>("/api-keys", body, options);
  }

  deleteApiKey<T = unknown>(keyId: string, options?: RequestOptions): Promise<T> {
    return this.http.delete<T>(`/api-keys/${encodeURIComponent(keyId)}`, options);
  }
}

File: src/api/api_keys.rs
use reqwest::Method;
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::{enc, Error, HttpClient};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum ApiKeyScope {
    Read,
    Write,
    Inference,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateApiKeyRequest {
    pub name: String,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub scopes: Vec<ApiKeyScope>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub allowed_ips: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub expires_at: Option<String>,
}

pub struct ApiKeysApi<'a> {
    http: &'a HttpClient,
}

impl<'a> ApiKeysApi<'a> {
    pub fn new(http: &'a HttpClient) -> Self {
        Self { http }
    }

    pub async fn list(&self) -> Result<Value, Error> {
        self.http
            .send(Method::GET, "/api-keys", &[], None, true)
            .await
    }

    pub async fn create(&self, request: &CreateApiKeyRequest) -> Result<Value, Error> {
        let body = serde_json::to_value(request).expect("API key request is serializable");
        self.http
            .send(Method::POST, "/api-keys", &[], Some(&body), true)
            .await
    }

    pub async fn revoke(&self, key_id: &str) -> Result<Value, Error> {
        let path = format!("/api-keys/{}", enc(key_id));
        self.http.send(Method::DELETE, &path, &[], None, true).await
    }
}
